File: 백준/Gold/17779. 게리맨더링 2/게리맨더링 2.java
import java.util.*;
import java.io.*;
public class Main {
    static int N;
    static int[][] board;
    static int answer;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        board = new int[N][N];
        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        answer = Integer.MAX_VALUE;

//        func(3, 2, 1, 1);

        for (int i = 0 ; i < N; i++) {
            for (int j = 0; j < N; j++) {
                for (int d1 = 1; d1 < N; d1++) {
                    for (int d2 = 1; d2 < N; d2++) {
                        if ((i + d1 + d2 < N) && (j - d1 >= 0) && (j + d2 < N)) {
                            answer = Integer.min(func(i, j, d1, d2), answer);
                        }
                    }
                }
            }
        }
        System.out.println(answer);
    }

    public static int func(int r, int c, int d1, int d2) {
        int max = Integer.MIN_VALUE;
        int min = Integer.MAX_VALUE;
        int[][] visited = new int[N][N];
        int[] tmp = new int[5];

        // 1번 경계선 긋기
        int dr = 0;
        int dc = 0;
        while (dr <= d1 && dc >= -d1) {
            visited[r+dr][c+dc] = 5;
            dr++;
            dc--;
        }

        // 2번 경계선 긋기
        dr = 0;
        dc = 0;
        while (dr <= d2 && dc <= d2) {
            visited[r+dr][c+dc] = 5;
            dr++;
            dc++;
        }

        // 3번 경계선 긋기
        dr = 0;
        dc = 0;
        while (dr <= d2 && dc <= d2) {
            visited[r+d1+dr][c-d1+dc] = 5;
            dr++;
            dc++;
        }

        // 4번 경계선 긋기
        dr = 0;
        dc = 0;
        while (dr <= d1 && dc >= -d1) {
            visited[r+d2+dr][c+d2+dc] = 5;
            dr++;
            dc--;
        }

        // 1번 선거구 계산하기
        for (int i = 0; i < r + d1; i++) {
            for (int j = 0; j <= c; j++) {
                if (visited[i][j] == 5) {
                    break;
                }
                visited[i][j] = 1;
                tmp[0] += board[i][j];
            }
        }

        // 2번 선거구 계산하기
        for (int i = 0; i <= r + d2; i++) {
            for (int j = N-1; j > c; j--) {
                if (visited[i][j] == 5) {
                    break;
                }
                visited[i][j] = 2;
                tmp[1] += board[i][j];
            }
        }

        // 3번 선거구 계산하기
        for (int i = r + d1; i < N; i++) {
            for (int j = 0; j < c - d1 + d2; j++) {
                if (visited[i][j] == 5) {
                    break;
                }
                visited[i][j] = 3;
                tmp[2] += board[i][j];
            }
        }

        // 4번 선거구 계산하기
        for (int i = r + d2 + 1; i < N; i++) {
            for (int j = N-1; j >= c - d1 + d2; j--) {
                if (visited[i][j] == 5) {
                    break;
                }
                visited[i][j] = 4;
                tmp[3] += board[i][j];
            }
        }

        // 5번 선거구 계산하기
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (visited[i][j] == 0 || visited[i][j] == 5) {
                    tmp[4] += board[i][j];
                }

            }
        }
        for (int i = 0; i < 5; i++) {
            if (tmp[i] > max) {
                max = tmp[i];
            }
            if (tmp[i] < min) {
                min = tmp[i];
            }
        }

        return max - min;
    }
}

File: 백준/Gold/17779. 게리맨더링 2/게리맨더링 2.py
import sys
input = sys.stdin.readline
N = int(input())
board = [list(map(int, input().split())) for _ in range(N)]

answer = 1e9

def func(r, c, d1, d2):
    global answer
    visited = [[0] * N for _ in range(N)]
    # 1번 경계선 긋기
    dr, dc = 0, 0
    while dr <= d1 and dc >= -d1:
        visited[r + dr][c + dc] = 5
        dr += 1
        dc -= 1

    # 2번 경계선 긋기
    dr, dc = 0, 0
    while dr <= d2 and dc <= d2:
        visited[r + dr][c + dc] = 5
        dr += 1
        dc += 1

    # 3번 경계선 긋기
    dr, dc = 0, 0
    while dr <= d2 and dc <= d2:
        visited[r + d1 + dr][c - d1 + dc] = 5
        dr += 1
        dc += 1

    # 4번 경계선 긋기
    dr, dc = 0, 0
    while dr <= d1 and dc >= -d1:
        visited[r + d2 + dr][c + d2 + dc] = 5
        dr += 1
        dc -= 1

    tmp = [0] * 5

    # 1번 선거구 인원수 세기
    for i in range(r+d1):
        for j in range(c+1):
            if visited[i][j] == 5:
                break
            visited[i][j] = 1
            tmp[0] += board[i][j]

    # 2번 선거구 인원수 세기
    for j in range(c+1, N):
        for i in range(r + d2+1):
            if visited[i][j] == 5:
                break
            visited[i][j] = 2
            tmp[1] += board[i][j]

    # 3번 선거구 인원수 세기
    for i in range(r + d1, N):
        for j in range(c - d1 + d2):
            if visited[i][j] == 5:
                break
            visited[i][j] = 3
            tmp[2] += board[i][j]

    # 4번 선거구 인원수 세기
    for i in range(r + d2 + 1, N):
        for j in range(N-1, c - d1 + d2 - 1, -1):
            if visited[i][j] == 5:
                break
            visited[i][j] = 4
            tmp[3] += board[i][j]

    for i in range(N):
        for j in range(N):
            if visited[i][j] == 0 or visited[i][j] == 5:
                tmp[4] += board[i][j]

    diff = max(tmp) - min(tmp)
    answer = min(diff, answer)


for i in range(N): # 행
    for j in range(N): # 열
        for d1 in range(1, N):
            for d2 in range(1, N):
                if (j - d1) >= 0 and (j + d2) < N and (i + d1 + d2) < N:
                    func(i, j, d1, d2)

# func(3, 2, 1, 1)
print(answer)

'''
7
1 2 3 4 1 6 1
7 8 9 1 4 2 2
2 3 4 1 1 3 3
6 6 6 6 9 4 4
9 1 9 1 9 5 5
1 1 1 1 9 9 6
1 1 1 1 9 9 6
'''